src/llvm-alloc-opt.cpp

```diff
@@ -0,0 +1,667 @@
+        return false;
+    };
+    if (simple_replace(orig_inst, new_inst))
+        return;
+    assert(stack.empty());
+    ReplaceUsesFrame cur{orig_inst, new_inst};
+    auto finish_cur = [&] () {
+        assert(cur.orig_i->user_empty());
+        if (cur.orig_i != orig_inst) {
+            cur.orig_i->eraseFromParent();
+        }
+    };
+    auto push_frame = [&] (Instruction *orig_i, Instruction *new_i) {
+        if (simple_replace(orig_i, new_i))
+            return;
+        stack.push_back(cur);
+        cur = {orig_i, new_i};
+    };
+    // Both `orig_i` and `new_i` should be pointer of the same type
+    // but possibly different address spaces. `new_i` is always in addrspace 0.
+    auto replace_inst = [&] (Instruction *user) {
+        Instruction *orig_i = cur.orig_i;
+        Instruction *new_i = cur.new_i;
+        if (isa<LoadInst>(user) || isa<StoreInst>(user)) {
+            user->replaceUsesOfWith(orig_i, new_i);
+        }
+        else if (auto call = dyn_cast<CallInst>(user)) {
+            if (ptr_from_objref && ptr_from_objref == call->getCalledFunction()) {
+                call->replaceAllUsesWith(new_i);
+                call->eraseFromParent();
+                return;
+            }
+            // remove from operand bundle
+            Type *new_t = new_i->getType();
+            user->replaceUsesOfWith(orig_i, ConstantPointerNull::get(cast<PointerType>(new_t)));
+        }
+        else if (isa<AddrSpaceCastInst>(user) || isa<BitCastInst>(user)) {
+            auto cast_t = PointerType::get(cast<PointerType>(user->getType())->getElementType(),
+                                           0);
+            auto replace_i = new_i;
+            Type *new_t = new_i->getType();
+            if (cast_t != new_t) {
+                replace_i = new BitCastInst(replace_i, cast_t, "", user);
+                replace_i->setDebugLoc(user->getDebugLoc());
+                replace_i->takeName(user);
+            }
+            push_frame(user, replace_i);
+        }
+        else if (auto gep = dyn_cast<GetElementPtrInst>(user)) {
+            SmallVector<Value *, 4> IdxOperands(gep->idx_begin(), gep->idx_end());
+            auto new_gep = GetElementPtrInst::Create(gep->getSourceElementType(),
+                                                     new_i, IdxOperands,
+                                                     gep->getName(), gep);
+            new_gep->setIsInBounds(gep->isInBounds());
+            new_gep->takeName(gep);
+            copyMetadata(new_gep, gep);
+            push_frame(gep, new_gep);
+        }
+        else {
+            abort();
+        }
+    };
+
+    while (true) {
+        replace_inst(cast<Instruction>(*cur.orig_i->user_begin()));
+        while (cur.orig_i->use_empty()) {
+            finish_cur();
+            if (stack.empty())
+                return;
+            cur = stack.back();
+            stack.pop_back();
+        }
+    }
+}
+
+bool AllocOpt::isSafepoint(Instruction *inst)
```

Comment:
Also note that when the lowering pass recognize more functions as non safepoint, we don't necessarily want to update this pass to include those. Here I included the same list just to be safe. In principle, this needs a list of functions that codegen assums to be not safepoint. There can be functions that are never safepoint but as long as neither codegen or llvm can insert a call into a unsafe use chain it doesn't need to be treated as non safepoint here.